sdk/communication/AzureCommunication/Source/Authentication/HMACAuthenticationPolicy.swift

```diff
@@ -0,0 +1,144 @@
+// rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+// sell copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+// IN THE SOFTWARE.
+//
+// --------------------------------------------------------------------------
+
+#if canImport(AzureCore)
+import AzureCore
+#endif
+
+import Foundation
+import CommonCrypto
+import CryptoKit
+
+@objcMembers public class HMACAuthenticationPolicy: Authenticating {
+    public var next: PipelineStage?
+    private let accessKey: String
+    
+    static let dateHeader = "date"
+    static let hostHeader  = "host"
+    static let contentHashHeader = "x-ms-content-sha256"
+    
+    public init(accessKey: String) {
+        self.accessKey = accessKey
+    }
+
+    public func authenticate(
+        request: PipelineRequest,
+        completionHandler: @escaping OnRequestCompletionHandler) {
+        let contents = request.httpRequest.data ?? Data() // Is this the body of the request?
+        
+        guard request.httpRequest.url.scheme?.contains("https") == true else {
+            completionHandler(
+                request,
+                AzureError.sdk("HMACAuthenticationPolicy requires a URL using the HTTPS protocol scheme"))
+            return
+        }
+        
+        request.httpRequest.headers = addAuthenticationHeaders(
+            url: request.httpRequest.url,
+            httpMethod: request.httpRequest.httpMethod.rawValue,
+            contents: contents)
+        
+        completionHandler(request, nil)
+    }
+    
+    public func addAuthenticationHeaders(
+        url: URL,
+        httpMethod: String,
+        contents: Data) -> HTTPHeaders {
+        var headers: HTTPHeaders = [:]
+        headers[HMACAuthenticationPolicy.contentHashHeader] = contents.sha256
+        
+        let dateFormatter = DateFormatter()
+        dateFormatter.dateFormat = "E, dd MMM YYYY HH:mm:ss 'GMT'" // Is this the right date format? Try ISO8601DateFormatter
+        dateFormatter.timeZone = TimeZone(abbreviation: "UTC")
+        
+        let date = Date()
+        let utcNow = dateFormatter.string(from: date)
+        headers[HMACAuthenticationPolicy.dateHeader] = utcNow
+        headers[HMACAuthenticationPolicy.hostHeader] = url.host
```

Comment:
can these be set in addSignatureHeader() as well? so that we don't  need to do `headers.merge`?

Reply:
Yes, but this feels a little backwards, would it make more sense to get rid of the `addSignatureHeader` method and add its implementation in here instead?